### PR TITLE
WIP: Sparse layers and (experimental) CFF2 support

### PR DESCRIPTION
**Pull request by anthrotype**
PTAL @madig

### CODE REVIEW

**Comment by anthrotype** [comment]
@m4rc1e Hey Marc, would you like to try installing fontmake from this PR branch and run it on some of the GF projects that use Glyphs "brace layer" feature?

Just make a new virtual env and run this pip command:

```
$ pip install -e git+https://github.com/anthrotype/fontmake.git@sparse-layers-and-cff2-support#egg=fontmake
```

Thanks!

/cc @davelab6

**Comment by davelab6** [comment]
Love it!! @mjlagattuta I guess Marc maybe busy so perhaps you can try this
on hepta

**Comment by anthrotype** [comment]
Yes, varLib expects masters to have unique locations whereas that glyphs file from which you are building contains duplicate brace layers, glyphsLib emits a warning. The user needs to fix the source.

**Comment by madig** [comment]
Note to self: try to make the assertion error message helpful.

Trying this on one of our commercial fonts that I fudged into VF shape. I'm trying to track down this error:

```
INFO:fontTools.varLib:Generating MVAR
Traceback (most recent call last):
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Programs\Python\Python37\lib\pdb.py", line 1695, in main
    pdb._runmodule(mainpyfile)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Programs\Python\Python37\lib\pdb.py", line 1540, in _runmodule
    self.run(code)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Programs\Python\Python37\lib\bdb.py", line 585, in run
    exec(cmd, globals, locals)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Temp\venv\lib\site-packages\fontmake\__main__.py", line 15, in <module>
    from __future__ import print_function, absolute_import
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Temp\venv\lib\site-packages\fontmake\__main__.py", line 284, in main
    project.run_from_designspace(designspace_path, **args)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Temp\venv\lib\site-packages\fontmake\font_project.py", line 821, in run_from_designspace
    **kwargs
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Temp\venv\lib\site-packages\fontmake\font_project.py", line 880, in _run_from_designspace_interpolatable
    ttf_designspace, output_path=output_path, output_dir=output_dir
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Temp\venv\lib\site-packages\fontmake\font_project.py", line 365, in build_variable_font
    font, _, _ = varLib.build(designspace, finder)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Temp\venv\lib\site-packages\fontTools\varLib\__init__.py", line 807, in build
    _add_MVAR(vf, model, master_fonts, axisTags)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Temp\venv\lib\site-packages\fontTools\varLib\__init__.py", line 550, in _add_MVAR
    base, varIdx = store_builder.storeMasters(master_values)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Temp\venv\lib\site-packages\fontTools\varLib\varStore.py", line 88, in storeMasters
    deltas = self._model.getDeltas(master_values)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Temp\venv\lib\site-packages\fontTools\varLib\models.py", line 359, in getDeltas
    assert len(masterValues) == len(self.deltaWeights)
```

**Comment by madig** [comment]
With glyphsLib git master and https://github.com/fonttools/fonttools/pull/1477, I can successfully build WorkSans and WorkSans-Italic (git master) as VF TTFs.

**Comment by madig** [comment]
Trying to build WorkSans-Italic into a variable-cff2 results in

```
Traceback (most recent call last):
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Temp\venv\Scripts\fontmake-script.py", line 11, in <module>
    load_entry_point('fontmake', 'console_scripts', 'fontmake')()
  File "c:\users\nikolaus.waxweiler\appdata\local\temp\venv\src\fontmake\lib\fontmake\__main__.py", line 276, in main
    project.run_from_glyphs(glyphs_path, **args)
  File "c:\users\nikolaus.waxweiler\appdata\local\temp\venv\src\fontmake\lib\fontmake\font_project.py", line 701, in run_from_glyphs
    self.run_from_designspace(designspace_path, **kwargs)
  File "c:\users\nikolaus.waxweiler\appdata\local\temp\venv\src\fontmake\lib\fontmake\font_project.py", line 821, in run_from_designspace
    **kwargs
  File "c:\users\nikolaus.waxweiler\appdata\local\temp\venv\src\fontmake\lib\fontmake\font_project.py", line 894, in _run_from_designspace_interpolatable
    ttf=False,
  File "c:\users\nikolaus.waxweiler\appdata\local\temp\venv\src\fontmake\lib\fontmake\font_project.py", line 365, in build_variable_font
    font, _, _ = varLib.build(designspace, finder)
  File "c:\userslocal\nikolaus.waxweiler\development\fonttools\lib\fontTools\varLib\__init__.py", line 785, in build
    vf = deepcopy(master_fonts[ds.base_idx])
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Programs\Python\Python37\lib\copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Programs\Python\Python37\lib\copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Programs\Python\Python37\lib\copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Programs\Python\Python37\lib\copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Programs\Python\Python37\lib\copy.py", line 150, in deepcopy
    y = copier(x, memo)
[...]
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Programs\Python\Python37\lib\copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "C:\Users\nikolaus.waxweiler\AppData\Local\Programs\Python\Python37\lib\copy.py", line 281, in _reconstruct
    if hasattr(y, '__setstate__'):
  File "c:\userslocal\nikolaus.waxweiler\development\fonttools\lib\fontTools\cffLib\__init__.py", line 2432, in __getattr__
    value = BaseDict.__getattr__(self, name)
  File "c:\userslocal\nikolaus.waxweiler\development\fonttools\lib\fontTools\cffLib\__init__.py", line 2245, in __getattr__
    value = self.rawDict.get(name, None)
  File "c:\userslocal\nikolaus.waxweiler\development\fonttools\lib\fontTools\cffLib\__init__.py", line 2432, in __getattr__
    value = BaseDict.__getattr__(self, name)
[...]
  File "c:\userslocal\nikolaus.waxweiler\development\fonttools\lib\fontTools\cffLib\__init__.py", line 2432, in __getattr__
    value = BaseDict.__getattr__(self, name)
  File "c:\userslocal\nikolaus.waxweiler\development\fonttools\lib\fontTools\cffLib\__init__.py", line 2245, in __getattr__
    value = self.rawDict.get(name, None)
RecursionError: maximum recursion depth exceeded while calling a Python object

**Comment by anthrotype** [comment]
CFF2 and sparse masters support is pending on https://github.com/fonttools/fonttools/pull/1475

**Comment by madig** [comment]
Might be worth holding off of offering the option until support exists then.

File: Lib/fontmake/font_project.py [hunk 1]
```diff
@@ -250,28 +260,109 @@ def build_ttfs(self, ufos, **kwargs):
...
+        conversion_error=None,
+        feature_writers=None,
+        cff_round_tolerance=None,
+        **kwargs
+    ):
+        if hasattr(designspace, "__fspath__"):
+            ds_path = designspace.__fspath__()
+        if isinstance(designspace, basestring):
+            ds_path = designspace
+        else:
```
Comment (madig) [inline Lib/fontmake/font_project.py]: Is it worth it to test `.path` for `None` and `raise` that fontmake shouldn't be used as a library?

Comment (madig) [reply Lib/fontmake/font_project.py]: Or maybe leave as is for now.

**Pull request merged**